cluster/gce/gci/nvidia-gpus/installer.sh

```diff
@@ -0,0 +1,207 @@
+    # git checkout "tags/v${kernel_version_stripped}"
+    git checkout ${LAKITU_KERNEL_SHA1}
+
+    # Prepare kernel configu and source for modules.
+    echo "Preparing kernel sources ..."
+    zcat "/proc/config.gz" > ".config"
+    make olddefconfig
+    make modules_prepare
+    # Done.
+    popd
+}
+
+download_install_nvidia() {
+    local pkg_name="${NVIDIA_CUDA_PKG_NAME}"
+    local url="${NVIDIA_CUDA_URL}"
+    local log_file_name="${NVIDIA_DRIVER_DIR}/nvidia-installer.log"
+
+    mkdir -p "${NVIDIA_DRIVER_DIR}"
+    pushd "${NVIDIA_DRIVER_DIR}"
+
+    echo "Downloading Nvidia CUDA package from ${url} ..."
+    curl -L -s "${url}" -o "${pkg_name}"
+    echo "${NVIDIA_CUDA_MD5SUM} ${pkg_name}" | md5sum --check
+
+    echo "Extracting Nvidia CUDA package ..."
+    sh ${pkg_name} --extract="$(pwd)"
+
+    echo "Running the Nvidia driver installer ..."
+    if ! sh "${NVIDIA_DRIVER_PKG_NAME}" --kernel-source-path="${KERNEL_SRC_DIR}" --silent --accept-license --keep --log-file-name="${log_file_name}"; then
+        echo "Nvidia installer failed, log below:"
+        echo "==================================="
+        tail -50 "${log_file_name}"
+        echo "==================================="
+        exit 1
+    fi
+    # Create unified memory device file.
+    nvidia-modprobe -c0 -u
+    popd
+}
+
+unlock_loadpin_and_reboot_if_needed() {
+    kernel_cmdline="$(cat /proc/cmdline)"
+    if echo "${kernel_cmdline}" | grep -q -v "lsm.module_locking=0"; then
+        local -r esp_partition="/dev/sda12"
+        local -r mount_path="/tmp/esp"
+        local -r grub_cfg="efi/boot/grub.cfg"
+
+        mkdir -p "${mount_path}"
+        mount "${esp_partition}" "${mount_path}"
+
+        pushd "${mount_path}"
+        cp "${grub_cfg}" "${grub_cfg}.orig"
+        sed 's/cros_efi/cros_efi lsm.module_locking=0/g' -i "efi/boot/grub.cfg"
+        cat "${grub_cfg}"
+        popd
+        sync
+        umount "${mount_path}"
+        # Restart the node for loadpin to be disabled.
+        echo b > /sysrq
+    fi
+}
+
+create_uvm_device() {
+    # Create unified memory device file.
+    nvidia-modprobe -c0 -u
+}
+
+verify_base_image() {
+    mount --bind /rootfs/etc/os-release /etc/os-release
+    local id="$(grep "^ID=" /etc/os-release)"
+    if [[ "${id#*=}" != "cos" ]]; then
+        echo "This installer is designed to run on Container-Optimized OS only"
+        exit 1
+    fi
+}
+
+setup_overlay_mounts() {
+    mkdir -p ${USR_WRITABLE_DIR} ${USR_WORK_DIR} ${LIB_WRITABLE_DIR} ${LIB_WORK_DIR} 
+    mount -t overlay -o lowerdir=/usr,upperdir=${USR_WRITABLE_DIR},workdir=${USR_WORK_DIR} none /usr
+    mount -t overlay -o lowerdir=/lib,upperdir=${LIB_WRITABLE_DIR},workdir=${LIB_WORK_DIR} none /lib
+}
+
+exit_if_install_not_needed() {
+    if nvidia-smi; then
+	echo "nvidia drivers already installed. Skipping installation"
+        post_installation_sequence
+	exit 0
+    fi
+}
+
+restart_kubelet() {
+    echo "Sending SIGTERM to kubelet"
```

Comment:
This is necessary for the kubelet to pick up the GPUs. Kubelet cannot support hotplugging GPUs for various reasons. We tried that with just PCI based data and it's proving to be hard.
Ideally, we need to reboot the entire node, but we haven't gotten there yet.